Initial work on `noosphere-p2p`, the libp2p-based DHT underpinings of a name system.

jsantell opened the pull request.

cdata commented:
Very exciting! I'll take a look at this starting today, although it might take me until tomorrow to get through it.

jsantell commented on rust/noosphere-p2p/src/dht/config.rs:
```diff
@@ -0,0 +1,178 @@
+}
+
+impl DHTConfig {
+    pub fn bootstrap_interval(mut self, interval: u64) -> Self {
```
These setters should be `&mut self` for easier non-one-liner configs

jsantell replied:
I take that back, as the config object is consumed by `DHTNode`.

jsantell commented on rust/noosphere-p2p/src/dht/node.rs:
```diff
@@ -0,0 +1,180 @@
+
+    /// Spawns a networking thread with a DHT node, returning the
+    /// corresponding [DHTNode].
+    pub fn spawn(config: DHTConfig) -> Result<DHTNode, DHTError> {
```
This could probably just be removed in lieu of `new`, but did not want to trivialize starting up a listener for a network

jsantell commented on rust/noosphere-p2p/src/dht/processor.rs:
```diff
@@ -0,0 +1,582 @@
+}
+
+/*
+macro_rules! take_requests {
```
Will remove

jsantell commented on rust/noosphere-p2p/src/dht/processor.rs:
```diff
@@ -0,0 +1,582 @@
+
+macro_rules! dht_trace {
+    ($self:expr, $event:expr) => {
+        trace!("{:#?}::{:#?}", $self.peer_id.to_base58(), $event);
```
Tracing in this event handler here is kind of a mess currently

jsantell commented on rust/noosphere-p2p/src/dht/processor.rs:
```diff
@@ -0,0 +1,582 @@
+    }
+}
+
+impl Drop for DHTProcessor {
```
Will remove

jsantell commented on rust/noosphere-p2p/Cargo.toml:
```diff
@@ -0,0 +1,23 @@
+tokio = { version = "1.15", features = ["sync", "io-util", "io-std", "macros", "rt", "rt-multi-thread"] }
+ucan = { version = "0.7.0-alpha.1" }
+ucan-key-support = { version = "0.7.0-alpha.1" }
+libp2p = { version = "0.48.0", default-features = true, features = [ "identify", "dns-tokio", "tcp-tokio", "websocket", "noise", "mplex", "yamux", "kad" ] }
```
TODO: Try with `default-features = false`

jsantell commented:
Revisions:
* Cleaned up the cargo manifests:
  * Remove the deps for the previous iteration of "name-system" (CLI, noosphere-cli stuff)
  * Updating libp2p from 0.48.0 to 0.49.0! There are a handful of feature flag changes and deprecated network behaviours in the latest release, so this cleans up some TODOs and flags.
  * Turn off `default-features` for libp2p.
* Update tests:
  * Skip integration tests in wasm builds.
  * Use real IP4 loopback addresses (`/ip4/127.0.0.1/tcp/42533`) instead of memory addresses (`/memory/123412341234`) for testing.
* API stuff:
  * Changed `DHTNode::spawn` to `DHTNode::new` -- it's attractive to use an intentionally different name to indicate some sort of non-trivial instantiation is occurring (e.g. a networking thread), but preferring the idioms and good docs for now.
  * `DHTConfig`: due to this struct containing immutable state that's used in multiple threads, in lieu of a builder pattern, I have setter methods on `DHTConfig` in a [consuming-builder style](https://rust-lang.github.io/api-guidelines/type-safety.html#consuming-builders). Anything I should change here for consistency throughout Noosphere?
  * Changed some `Option<T>`s to `T` in `DHTNode`; as the network starts upon instantiation (`DHTNode::new`) and is useless after termination (e.g., restarting means creating a new instance).
* Removed some garbage

cdata commented on rust/noosphere-p2p/src/dht/transport.rs:
```diff
@@ -0,0 +1,35 @@
+        let dns_tcp = dns::TokioDnsConfig::system(tcp::TokioTcpTransport::new(
+            tcp::GenTcpConfig::new().nodelay(true),
+        ))?;
+        let ws_dns_tcp = websocket::WsConfig::new(dns::TokioDnsConfig::system(
```
Just curious: does the WebSocket transport help for cases beyond peering with browser nodes?

jsantell replied:
Probably can be removed at least for now, a peer can offer multiple addresses at once, and unclear to me yet how the web version works in context

cdata commented on rust/noosphere-p2p/src/dht/types.rs:
```diff
@@ -0,0 +1,130 @@
+#[derive(Debug)]
+pub enum DHTRequest {
+    Bootstrap,
+    //WaitForPeers(usize),
```
Is this for the case where you want to ensure a peer quorum of some size?

jsantell replied:
In lieu of "events" (messages from processor to client), this was a work around to get tests just running with some form of confidence. I think "events" would remove the need for this (with maybe the client side still having a helper function for it, maybe just in tests, built off of said Events). Other DHT commands like putting records can have a quorum constraint as well

cdata commented on rust/noosphere-p2p/tests/integration_test.rs:
```diff
@@ -0,0 +1,103 @@
+
+    for info in swarm_command(&mut client_nodes, |c| c.network_info()).await? {
+        assert_eq!(info.num_peers, num_clients);
+        // @TODO the number of connections seem inconsistent??
```
Same feedback re: `TODO` comments above

cdata reviewed (changes requested):
This is awesome. I was able to check it out and build it locally, while is also awesome :sweat_smile: 

My feedback is fairly minor. I think this is a great start and we can move forward from this foundation.

jsantell commented on rust/noosphere-p2p/src/dht/behaviour.rs:
```diff
@@ -0,0 +1,75 @@
+
+pub type DHTSwarmEvent = SwarmEvent<
+            <DHTBehaviour as swarm::NetworkBehaviour>::OutEvent,
+            <<<DHTBehaviour as swarm::NetworkBehaviour>::ConnectionHandler as IntoConnectionHandler>::Handler as ConnectionHandler>::Error>;
```
Yeah :cry: this SwarmEvent type expanded once introducing a second "sub behaviour" (Identify)

cdata commented on rust/noosphere-p2p/src/dht/config.rs:
```diff
@@ -0,0 +1,107 @@
+    /// during bootstrap. A standalone bootstrap node would
+    /// have this field empty.
+    pub bootstrap_peers: Vec<libp2p::Multiaddr>,
+    pub keypair: libp2p::identity::Keypair,
```
As mentioned elsewhere, the rule of thumb for me is that if we are defining the interface, we should err on the side of constructs that we already use elsewhere in our own interfaces. It's not always an option, but at least in the case of keys we have a thing we use everywhere.

jsantell commented on rust/noosphere-p2p/src/dht/processor.rs:
```diff
@@ -0,0 +1,483 @@
+                    num_provider_peers: _,
+                } => {}
+                kad::InboundRequest::AddProvider { record: _ } => {}
+                kad::InboundRequest::GetRecord {
```
Actually, the VPS/e2e tests show GetRecord can fetch the data from peers without a local cache, but doesn't illustrate a deeper network "search"

jsantell commented on rust/noosphere-p2p/src/dht/behaviour.rs:
```diff
@@ -0,0 +1,76 @@
+            <DHTBehaviour as swarm::NetworkBehaviour>::OutEvent,
+            <<<DHTBehaviour as swarm::NetworkBehaviour>::ConnectionHandler as IntoConnectionHandler>::Handler as ConnectionHandler>::Error>;
+
+#[derive(NetworkBehaviour)]
```
Adding this entry fixed it!
```
"libp2p_swarm_derive": [
      "NetworkBehaviour"
],
```

jsantell replied:
Nevermind, it did not :cry:

jsantell commented on rust/noosphere-p2p/src/dht/behaviour.rs:
```diff
@@ -0,0 +1,76 @@
+use crate::dht::DHTConfig;
+use libp2p::{
+    // In the next version of libp2p-identify 0.40.0 (the subpackage), these values
```
Out of date, now that we're using libp2p 0.49.0 here

jsantell commented:
Updated based off of feedback; noteworthy changes:

* Added `DHTNodeBuilder` to offer a more noosphere-esque interface rather than libp2p (e.g. `Ed25519KeyMaterial` and `String` instead of `libp2p::Keypair` and `libp2p::Multiaddr` and provide an option that doesn't eagerly generate a key.
  * `DHTConfig::default` remains mostly for testing. The non-builder creation methods (`DHTNode::new`) are still public, specifically for integration tests. Could move `DHTConfig::default` exclusively to tests.
  *  The `DHTNodeBuilder` example docs and integration tests illustrate the challenges of using non-libp2p objects as an interface, when getters/computed values are needed to populate other nodes (e.g. connecting multiple bootstrap nodes to each other, requiring addresses for each to be shared with each node). Mostly an issue with `Multiaddr` properties (`listening_address`, `bootstrap_peers`). The public interface of `DHTNode` could convert these internal `Multiaddr`s back to String, but that would ultimately get converted back into a `Multiaddr` at some point. As an alternative to `String` interfaces, essentially a typeless `Multiaddr`, we could introduce a different type, like `TcpSocket` and `Ipv4Addr`, but that same translation will still need to be done when wiring multiple nodes together.
* Remove web socket transport

cdata reviewed (approved):
This is a very exciting start to something extremely cool :beers: awesome work @jsantell !
